Add SPI Library

Ayush1325 opened the pull request:
- Tested master mode on beagleconnect freedom.
- Add master sample.
- Interrupts not implemented. Slave mode not tested.
- Fixes #87

Ayush1325 committed:
libraries: SPI: Initial SPI implementation

- Tested controller mode on beagleconnect_freedom

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 committed:
beagleconnect_freedom: Add SPI support

- Only supports master mode

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
samples: Add spi master sample

- Simple sample that sends bytes to SPI slave
- Tested on beagleconnect freedom

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

DhruvaG2000 commented on libraries/SPI/SPI.cpp:
```diff
@@ -0,0 +1,101 @@
+#include "SPI.h"
+#include <zephyr/kernel.h>
+
+uint8_t SPCR = 0;
```
No need of ``= 0`` IMO
See https://stackoverflow.com/questions/16015656/are-global-variables-always-initialized-to-zero-in-c

DhruvaG2000 replied:
Also, may want to add a comment at the top saying
```
/* Serial Peripheral Control Register */
```
As some people may even call it SPIGCR or something else. 
See https://ti.com/lit/ug/sprugp2a/sprugp2a for eg.

Ayush1325 replied:
@DhruvaG2000 the name is taken from [Arduino SPI docs](https://docs.arduino.cc/tutorials/generic/introduction-to-the-serial-peripheral-interface/). So I think we should stick with it?

DhruvaG2000 replied:
@Ayush1325  My comment wasn't on the name, I'm just saying write the full form as a comment above

DhruvaG2000 commented on libraries/SPI/SPI.cpp:
```diff
@@ -0,0 +1,101 @@
+      .buffers = &rx_buf,
+      .count = 1,
+  };
+  int ret;
```
No mixed declarations please.
FYI: https://langdev.stackexchange.com/questions/3115/why-are-mixed-declarations-more-challenging-to-implement-than-forcing-all-declar

Ayush1325 replied:
@DhruvaG2000 Do you just mean moving `ret` at the top or remove mixed declaration (like C89)? That would require making everything non-const.

DhruvaG2000 replied:
just move it to the top.

DhruvaG2000 commented on libraries/SPI/SPI.cpp:
```diff
@@ -0,0 +1,101 @@
+  };
+  int ret;
+
+
```
remove extra blank lines

DhruvaG2000 commented on samples/spi_master/README.rst:
```diff
@@ -0,0 +1,9 @@
+.. _hello_world:
+
+SPI Master
```
Please make use of more inclusive terminology.  This should be addressed throughout this PR.
Read https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/

SPI controller makes more sense.

DhruvaG2000 commented on samples/spi_master/src/app.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "SPI.h"
+#include <Arduino.h>
+
+#define SS 3
```
Very ambiguous name, use something more verbose

Ayush1325 replied:
So, the reason for choosing `SS` is that it seems to be defined in standard arduino as the CS Pin.

DhruvaG2000 replied:
https://docs.arduino.cc/tutorials/generic/introduction-to-the-serial-peripheral-interface/

chip sel/ CS makes more sense.

DhruvaG2000 replied:
> Controller/peripheral is formerly known as master/slave. Arduino no longer supports the use of this terminology. See the table below to understand the new terminology

Ayush1325 committed:
Interrupt support

- Disable registered interrupts on beginTransaction and re-enable on
  endTransaction.
- Add zephyrInternal.h for APIs not meant for external consumption

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

DhruvaG2000 reviewed (approved):
I think we'll also have to add the spi node to add support for other platforms as well, but I don't suppose you'd have the hw to test it all anyway.
So I am good with this PR, thanks for addressing all concerns!
